File: apps/web/src/utils/counterArrivalStore.ts
/**
 * counterArrivalStore.ts — Shared Counters P3 last-seen persistence (web).
 *
 * Per-board `taskId → displayed` snapshot, persisted in `localStorage`. This is
 * NOT synced schema — the arrival banner is a device-local "since you last
 * looked at THIS board" signal, so it deliberately lives outside the
 * board/task records that round-trip through Firestore. Mirrors the iOS
 * `CounterArrivalStore` (UserDefaults); both platforms key the entry by the
 * board id.
 *
 * All access is wrapped in try/catch — a disabled or full `localStorage` must
 * degrade to "no baseline" (⇒ first-view semantics, never a crash), because
 * the banner is strictly best-effort.
 */

/** Namespaced key prefix. The board id is appended verbatim. */
const KEY_PREFIX = 'oybc.counterArrivals.lastSeen.';

function keyFor(boardId: string): string {
  return `${KEY_PREFIX}${boardId}`;
}

/**
 * Read the last-seen snapshot for a board.
 *
 * @param boardId - The board whose snapshot to read.
 * @returns `taskId → displayed` map; `{}` when absent, malformed, or storage
 *   is unavailable (an absent entry is a first view — never an arrival).
 */
export function readLastSeen(boardId: string): Record<string, number> {
  try {
    const raw = localStorage.getItem(keyFor(boardId));
    if (!raw) return {};
    const parsed: unknown = JSON.parse(raw);
    if (!parsed || typeof parsed !== 'object') return {};
    const out: Record<string, number> = {};
    for (const [k, v] of Object.entries(parsed as Record<string, unknown>)) {
      if (typeof v === 'number' && Number.isFinite(v)) out[k] = v;
    }
    return out;
  } catch {
    return {};
  }
}

/**
 * Persist the last-seen snapshot for a board.
 *
 * @param boardId - The board whose snapshot to write.
 * @param snapshot - `taskId → displayed` map (from `snapshotCounterSquares`).
 */
export function writeLastSeen(boardId: string, snapshot: Record<string, number>): void {
  try {
    localStorage.setItem(keyFor(boardId), JSON.stringify(snapshot));
  } catch {
    // Ignore quota / disabled-storage errors — the banner is best-effort.
  }
}

File: apps/ios/OYBC/Services/CounterArrivalStore.swift
import Foundation

/// Device-local last-seen snapshot store for the Shared Counters P3 arrival
/// banner.
///
/// Persists a per-board `taskId → displayed` map in `UserDefaults`. This is
/// deliberately NOT synced schema — the arrival banner is a "since you last
/// looked at THIS board on THIS device" signal, so it lives outside the
/// board/task records that round-trip through Firestore. Mirrors the web
/// `counterArrivalStore.ts` (`localStorage`); both platforms key the entry by
/// the board id.
///
/// House style follows `TutorialProgressStore` — `UserDefaults`-backed with an
/// injectable `defaults` so unit tests can pass an isolated suite. Unlike that
/// store this is a plain value-less service (no `@Published`): the arrival VM
/// reads/writes it directly, nothing observes it.
final class CounterArrivalStore {

    private let defaults: UserDefaults

    /// - Parameter defaults: Injected for tests; defaults to `.standard`.
    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    private func key(boardId: String) -> String {
        "counterArrivals.lastSeen.\(boardId)"
    }

    /// The last-seen snapshot for a board.
    ///
    /// - Parameter boardId: The board whose snapshot to read.
    /// - Returns: `taskId → displayed` map; `[:]` when absent or malformed (an
    ///   absent entry is a first view — never an arrival).
    func lastSeen(boardId: String) -> [String: Int] {
        guard let data = defaults.data(forKey: key(boardId: boardId)),
              let map = try? JSONDecoder().decode([String: Int].self, from: data)
        else { return [:] }
        return map
    }

    /// Persist the last-seen snapshot for a board.
    ///
    /// - Parameters:
    ///   - boardId: The board whose snapshot to write.
    ///   - snapshot: `taskId → displayed` map (from `snapshotCounterSquares`).
    func save(boardId: String, snapshot: [String: Int]) {
        guard let data = try? JSONEncoder().encode(snapshot) else { return }
        defaults.set(data, forKey: key(boardId: boardId))
    }
}
